src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+        hmacObj.init(new SecretKeySpec(salt, "HKDF-Salt"));
+
+        if (inputKey == null) {
+            return hmacObj.doFinal();
+        } else {
+            return hmacObj.doFinal(inputKey.getEncoded());
+        }
+    }
+
+    /**
+     * Perform the HMAC-Expand operation.  At the end of the operation, the
+     * keyStream instance variable will contain the complete KDF output based on
+     * the input values and desired length.
+     *
+     * @param prk
+     *     the pseudorandom key used for HKDF-Expand
+     * @param info
+     *     optional context and application specific information or {@code null}
+     *     if no info data is provided.
+     * @param outLen
+     *     the length in bytes of the required output
+     *
+     * @return a byte array containing the complete KDF output.  This will be at
+     *     least as long as the requested length in the {@code outLen}
+     *     parameter, but will be rounded up to the nearest multiple of the HMAC
+     *     output length.
+     *
+     * @throws InvalidKeyException
+     *     if an invalid key was provided through the {@code HkdfParameterSpec}
+     *     or derived during the generation of the PRK.
+     */
+    protected byte[] hkdfExpand(SecretKey prk, byte[] info, int outLen)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+        byte[] kdfOutput;
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+
+        // Calculate the number of rounds of HMAC that are needed to
+        // meet the requested data.  Then set up the buffers we will need.
+        hmacObj.init(prk);
+        if (info == null) {
+            info = new byte[0];
+        }
+        int rounds = (outLen + hmacLen - 1) / hmacLen;
+        kdfOutput = new byte[rounds * hmacLen];
+        int offset = 0;
+        int tLength = 0;
+
+        for (int i = 0; i < rounds; i++) {
+
+            // Calculate this round
+            try {
+                // Add T(i).  This will be an empty string on the first
+                // iteration since tLength starts at zero.  After the first
+                // iteration, tLength is changed to the HMAC length for the
+                // rest of the loop.
+                hmacObj.update(kdfOutput, Math.max(0, offset - hmacLen),
+                               tLength);
+                hmacObj.update(info);                       // Add info
+                hmacObj.update((byte) (i + 1));              // Add round number
+                hmacObj.doFinal(kdfOutput, offset);
+
+                tLength = hmacLen;
+                offset += hmacLen;                       // For next iteration
+            } catch (ShortBufferException sbe) {
+                // This really shouldn't happen given that we've
+                // sized the buffers to their largest possible size up-front,
+                // but just in case...
+                throw new RuntimeException(sbe);
+            }
+        }
```

Comment:
`tLength` may not be necessary. Variables only used inside the loop can also be moved to the loop.
The loop can be modified as:

```
for (int i = 0, offset = 0; i < rounds; i++, offset += hmacLen) {
            // Calculate this round
            try {
                if (i > 0) {
                    hmacObj.update(kdfOutput, offset - hmacLen, hmacLen); // add T(i-1)
                }
                hmacObj.update(info);                       // Add info
                hmacObj.update((byte) (i + 1));             // Add round number
                hmacObj.doFinal(kdfOutput, offset);
            } catch (ShortBufferException sbe) {
                // This really shouldn't happen given that we've
                // sized the buffers to their largest possible size up-front,
                // but just in case...
                throw new RuntimeException(sbe);
            }
        }

```

Reply:
Sure, I added an if-check to skip the `hmacObj.update(...)` call for the i=0 case. This way, you don't need to set `tLength` to 0 and then changing it to `hmacLen` afterwards. I find it easier to understand.

Reply:
You can double-check if output is the same.
Conceptually, if you call update() with a 0-length input vs skipping the update() call, they should be equivalent.